File: 0786-k-th-smallest-prime-fraction/0786-k-th-smallest-prime-fraction.py
class Solution(object):
    def find_position(self, arr, mid):
        j, total, num, den, n = 1, 0, 0, 0, len(arr)
        frac = 0.0

        for i in range(n):
            while j < n and arr[i] >= arr[j] * mid:
                j += 1
            total += n - j

            if j < n and frac < arr[i] * 1.0 / arr[j]:
                frac = arr[i] * 1.0 / arr[j]
                num, den = i, j

        return [total, num, den]
            
    def kthSmallestPrimeFraction(self, arr, k):
        n = len(arr)
        left, right = 0, 1.0
        ans = []

        while left <= right:
            mid = (right - left) / 2 + left
            pos = self.find_position(arr, mid)
            if pos[0] == k:
                ans = [arr[pos[1]], arr[pos[2]]]
                break

            if pos[0] > k:
                right = mid
            else:
                left = mid

        return ans
        

File: 0786-k-th-smallest-prime-fraction/0786-k-th-smallest-prime-fraction.java
class Solution {
    public int[] kthSmallestPrimeFraction(int[] arr, int k) {
        int n = arr.length;
        PriorityQueue<int[]> pq = new PriorityQueue<>(new Comparator<int[]>() {
            public int compare(int[] a, int[] b) {
                int x = arr[a[0]]*arr[b[1]];
                int y = arr[b[0]]*arr[a[1]];
                return x - y;
            }
        });

        for(int i = 0; i<n-1; i++) {
            pq.add(new int[] {i, n-1});
        }

        while(k > 1) {
            int[] t = pq.remove();
            if(t[1] - 1 > t[0]) {
               t[1]--;
               pq.add(t);
            }
            k--;
        }

        int[] t = pq.peek();
        return new int[] {arr[t[0]], arr[t[1]]};
    }
}